### PR TITLE
Populate new unit test properties and add flag to include unit tests in compiled graph

### PR DESCRIPTION
**Pull request by fernst**
Populated new properties in unit test proto during compilation.

Also added a new workflow config setting to include unit tests in compiled DAG.

### CODE REVIEW

File: core/actions/test.ts [hunk 1]
```diff
@@ -188,6 +202,28 @@ export class Test extends ActionBuilder<dataform.Test> {
       } else {
         const refReplacingContext = new RefReplacingContext(testContext);
         this.proto.testQuery = refReplacingContext.apply(dataset.contextableQuery);
+        this.testTarget = dataform.Target.create(dataset.getTarget());
+
+        // Set the test query with the fully qualified table references.
+        if (dataset instanceof Table) {
```
Comment (ikholopov-omni) [inline core/actions/test.ts]: These if/else branches are equivalent

**Review by ikholopov-omni — CHANGES_REQUESTED** [review]

File: core/actions/test.ts (hunk 1 above)
Comment (fernst) [reply core/actions/test.ts]: True. I've consolidated this code.

File: core/actions/test.ts [hunk 2]
```diff
@@ -188,6 +202,19 @@ export class Test extends ActionBuilder<dataform.Test> {
...
+
+        // Set the test query with the fully qualified table references.
+        this.proto.target = this.replaceTestNameInTarget(
+          testTarget.getTarget(),
+          this.proto.name
+        );
+        this.proto.canonicalTarget = this.replaceTestNameInTarget(
+          testTarget.getCanonicalTarget(),
+          this.proto.name
+        );
```
Comment (kolina) [inline core/actions/test.ts]: Do I understand correctly that this can change the result target of the test?

I imagine this can break in some corner cases (the first thing coming to my mind: if tests become regular DAG nodes, someone can explicitly add dependency on the test node and this dependency can break if dataset / database part changed).

Maybe we can have behaviour similar to assertion nodes here instead (AFAIK they don't have this logic)

File: core/actions/test_test.ts
Comment (kolina) [inline core/actions/test_test.ts:1]: I think it's worth adding a test where logic of calling `replaceTestNameInTarget` actually matters

**Review by kolina — CHANGES_REQUESTED** [review]
Can you please against the latest `main`? So integration tests will run with non-expired credentials

File: core/actions/test.ts (hunk 2 above)
Comment (fernst) [reply core/actions/test.ts]: The function naming is a bit confusing now that I reread it.

This simply takes the parent target/canonical target and sets the test name as the new action name.

This is so we can generate a whole new target for the current test, which ensures we can use this as a proper DAG element (where the target is the identifier).

File: core/actions/test_test.ts
Comment (fernst) [reply core/actions/test_test.ts:1]: I renamed the function so it's more clear that this is only impacting the current action.

I'll add a test where I have 2 targets with similar name, and test names, but different schema

`defaultProject.schema1.tbl` and `defaultProject.schema2.tbl`, both with a test named `tbl_test` for each one.

Comment (fernst) [reply core/actions/test_test.ts:1]: Actually, there is an issue with 2 actions with the same name (meaning, you can't write tests for an action if 2 actions share the same name). I will fix this in a follow-up PR.

File: core/actions/test.ts (hunk 2 above)
Comment (fernst) [reply core/actions/test.ts]: I've renamed this function, PTAL.

File: core/actions/test_test.ts
Comment (fernst) [reply core/actions/test_test.ts:1]: I added a test case with 2 targets using the same table name but under different schemas. This tests the resolution logic for the targets and that actions are labeled under the same database/schema as the action being tested.

File: core/actions/test.ts [hunk 3]
```diff
@@ -198,6 +226,14 @@ export class Test extends ActionBuilder<dataform.Test> {
       VerifyProtoErrorBehaviour.SUGGEST_REPORTING_TO_DATAFORM_TEAM
     );
   }
+
+  private overrideTargetWithNewName(target: dataform.ITarget, testName: string): dataform.Target {
```
Comment (ikholopov-omni) [inline core/actions/test.ts]: Nit: Can be a local function, as it doesn't depend on `this`.

File: core/session.ts [hunk 4]
```diff
@@ -549,11 +558,24 @@ export class Session {
...
+          .forEach(tableOrViewAction => tableOrViewAction.dependencies(utils.resolvableAsTarget(test.getTarget())));
+      });
+  }
+
   private fullyQualifyDependencies(actions: ActionProto[]) {
     actions.forEach(action => {
       const fullyQualifiedDependencies: { [name: string]: dataform.ITarget } = {};
       if (action instanceof dataform.Declaration || !action.dependencyTargets) {
-        // Declarations cannot have dependencies.
+        // Declarations cannot have dependencies.
```
Comment (ikholopov-omni) [inline core/session.ts]: Nit: dangling whitespace?

File: core/actions/index.ts [hunk 5]
```diff
@@ -19,15 +20,17 @@ export type Action =
   | Assertion
   | Declaration
   | Notebook
-  | DataPreparation;
+  | DataPreparation
+  | Test;
```
Comment (ikholopov-omni) [inline core/actions/index.ts:24]: The addition of tests to the list of targets in the compiled graph and a dependency, changes in session.test backing structure is technically a breaking change.

At the very least we should bump a minor version, once we merger theses changes.

**Review by ikholopov-omni — DISMISSED** [review]

File: core/actions/index.ts (hunk 5 above)
Comment (fernst) [reply core/actions/index.ts:24]: I've since removed the dependency changes (meaning tests are in the DAG, but not connected to any other action). 

Should we still do the minor version bump? Or can we keep it under patch versions until we add the new tests into the middle of the DAG?

File: core/actions/test.ts (hunk 3 above)
Comment (fernst) [reply core/actions/test.ts]: Moved.

File: core/session.ts (hunk 4 above)
Comment (fernst) [reply core/session.ts]: Removed

File: core/actions/index.ts (hunk 5 above)
Comment (ikholopov-omni) [reply core/actions/index.ts:24]: At the moment there are still breaking changes in Dataform Core API, specifically in Session object. Bottom line is these changes shouldn't be released under a patch, which means that we should either:
1) Exercise control and block all @dataform/core releases until all breaking changes are submitted from the feature branch
2) Prepare all breaking changes in a separate branch and them merge them in a single commit with a minor version bump.

At the moment we don't have any automation to enforce (1) repo-wide. I think executing (2) is easier.

Comment (kolina) [reply core/actions/index.ts:24]: Current changes LGTM from my side, will wait for resolution in this thread to approve the PR

**Review by kolina — COMMENTED** [review]
Waiting on resolution of how to update Dataform core version with these changes

File: core/actions/index.ts (hunk 5 above)
Comment (fernst) [reply core/actions/index.ts:24]: I've added a new configuration property to only add the unit tests to the compiled DAG when this setting is set. This ensures backwards compatibility and ensures the graph remains consistent for users who want to keep this behavior.

File: core/session.ts [hunk 6]
```diff
@@ -732,6 +743,19 @@ export class Session {
...
   }
 
+  private addTestsToCompiledGraph(actions: Action[]) {
+    actions
+      .filter(action => action instanceof Test)
+      .map(test => test as Test)
+      .forEach(test => {
+        this.indexedActions
+          .find(test.getTestTarget())
+          .filter(action => action instanceof Table || action instanceof View)
```
Comment (kolina) [inline core/session.ts:753]: should we throw an error otherwise?

**Review by kolina — APPROVED** [review]

File: core/session.ts (hunk 6 above)
Comment (fernst) [reply core/session.ts:753]: This would have failed already in https://github.com/dataform-co/dataform/blob/main/core/actions/test.ts#L180-L195 , as that already checks the type of the action that the test is testing.
